### ISSUE TITLE
`setVersion()` should also accept other version format, not just `x.y.z`. 

### ISSUE THREAD

**Issue by cowfox** [SamVerschueren/cordova-config#14]
Currently, `setVersion()` uses the regex check to only accept `x.y.z`. 

```
var regex = new RegExp('^[0-9]+\.[0-9]+\.[0-9]+$');
```

However, if based on [semver](http://semver.org/), it should also support **pre-release** version #, like `1.0.0-alpha.1`.

**Comment by SamVerschueren** [comment]
It's based upon the [documentation](https://cordova.apache.org/docs/en/6.0.0/config_ref/index.html#link-core-configuration-elements) of cordova itself.

> The `<widget>` element's `id` attribute provides the app's reverse-domain identifier, and the `version` its full version number expressed in major/minor/patch notation.

Do you have some docs where it states it accepts pre-release versions?

**Comment by cowfox** [comment]
Although in the final app submission, it only accepts `x.y.x` format, in the actual practice, it still gets chance to use a **pre-release** version #. 

After assigning `version` to a **pre-release** #, it can still build the app successfully. 

I would recommand to use `semver.valid()` to help validate the version #. - https://github.com/npm/node-semver

**Comment by SamVerschueren** [comment]
So you are able to sign the application with a pre-release version but you are not able to submit it to the store?

**Comment by SamVerschueren** [comment]
So you are able to sign the application with a pre-release version but you are not able to submit it to the store? And what about beta testing via the app store or testflight?

Have to check this first. I don't want people to be able to use it and then getting an error when they try to publish the application in beta or in release. I don't see the point of using a pre-release version if it fails in the store.

**Comment by cowfox** [comment]
As far as I know for the **iOS** development, 
- Since now **Testflight** has been integrated into **iTunes Connect** and all of the **beta** builds are organized by each of the "versions", it has to be following the official rule, which can be `x.y` or `x.y.z`. 
- However, when doing **In-house** app build (for both test and release), since it is handled per each organization, it can actually support a more types of version, like `x.y.x-beta.u`. 

Not sure about how Android works today.

**Comment by theJuddson** [comment]
I'm a little late here, but I've recently started leveraging this module in my Cordova build process. I too would like to see semver support for use in our internal pre-release builds. My suggestion would be to add an optional second argument (e.g. 'allowSemver' or 'relaxVersionCheck' or 'allowPreReleaseTags') to this function that could be used to relax the version checking to allow for pre-release tags. Then it's up to the developer to ensure that the versions they publish are valid for the environment they are publishing too. 

Just a suggestion. I've worked around the problem for our internal builds.

**Comment by SamVerschueren** [comment]
I agree. Don't have the time to work on this now but I'll accept a PR. If someone takes this, `allowSemver` would to unless someone comes up with something better.

**Comment by myrddraall** [comment]
I submitted pr #20 to address this

**Comment by piotr-cz** [comment]
I'd like to use semver for production builds, but for any other type I'd like to use output of `git describe`, as outlined in these articles:
- [Configuring Android Project — Version Name & Code](https://proandroiddev.com/configuring-android-project-version-name-code-b168952f3323)
- [Using git to generate versionName and versionCode for Android apps](https://blog.uncommon.is/using-git-to-generate-versionname-and-versioncode-for-android-apps-aaa9fc2c96af).

versionName could then be for example `1.0.0-beta.3-11-gcfb267d-dirty`

While it's impossible to come up with validation that would fit any developer need, I'm for a flag that would disable validation of versionName element

**Comment by piotr-cz** [comment]
In fact, SemVer 2.0.0 [allows appending build metadata](https://semver.org/spec/v2.0.0.html#spec-item-10) so the example would be `1.0.0-beta.3+11.gcfb267d.dirty`
